pkg/datasource/data.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datasource
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"slices"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+)
+
+type (
+	data  api.GadgetData
+	field api.Field
+)
+
+func (*data) private() {}
+
+func (f *field) ReflectType() reflect.Type {
+	switch Kind(f.Type) {
+	default:
+		return nil
+	case Int8:
+		return reflect.TypeOf(int8(0))
+	case Int16:
+		return reflect.TypeOf(int16(0))
+	case Int32:
+		return reflect.TypeOf(int32(0))
+	case Int64:
+		return reflect.TypeOf(int64(0))
+	case Uint8:
+		return reflect.TypeOf(uint8(0))
+	case Uint16:
+		return reflect.TypeOf(uint16(0))
+	case Uint32:
+		return reflect.TypeOf(uint32(0))
+	case Uint64:
+		return reflect.TypeOf(uint64(0))
+	case Float32:
+		return reflect.TypeOf(float32(0))
+	case Float64:
+		return reflect.TypeOf(float64(0))
+	case Bool:
+		return reflect.TypeOf(false)
+	case String:
+		return reflect.TypeOf("")
+	}
+}
```

Comment:
Why Array and Slice aren't considered here? I tried to add them but ig exploded.

Reply:
I think they (and String) need special handling because it's no simple conversion - if there's no reflect type, a field will be handled as binary blob / CString for now. A gadget author should be able to specify in more detail in the metadata file how the field should be encoded/decoded (so that JSON would have hex/base64 data instead of a string representation in some cases, for example). That's not a part of the PR right now - do you think it should be?